src/Psalm/Internal/Codebase/Populator.php

```diff
@@ -264,15 +285,29 @@ private function populateClassLikeStorage(ClassLikeStorage $storage, array $depe
             }
         }
 
-        $this->populateOverriddenMethods($storage);
+        $this->populateOverriddenMethods($storage, $storage_provider);
 
         $this->progress->debug('Have populated ' . $storage->name . "\n");
 
         $storage->populated = true;
 
         if (isset($this->invalid_class_storages[$fq_classlike_name_lc])) {
             foreach ($this->invalid_class_storages[$fq_classlike_name_lc] as $dependency) {
+                if (isset($dependency->dependent_classlikes[$fq_classlike_name_lc])) {
+                    if ($dependency->location) {
+                        IssueBuffer::maybeAdd(
+                            new CircularReference(
+                                'Circular reference discovered when loading ' . $dependency->name,
+                                $dependency->location
+                            )
+                        );
+                    }
+
+                    continue;
+                }
+
                 $dependency->populated = false;
+                unset($dependency->invalid_dependencies[$fq_classlike_name_lc]);
```

Comment:
This is the important part.

Reply:
So, if I get this part right, the stub reader will try to read a stub, stumbles upon an unknown dependency, so it will add it in invalid_dependencies and invalid_class_storages. Then when it's done, it loops over invalid_class_storages again to check if stubs read later found the dependency. If you find a circular dependency like A -> B and B -> A, you emit an issue.

So what about A -> B, B -> C, C -> A? Would that be detected?

Reply:
I think so, but I'll update the test.

Reply:
I guess it doesn't, I'll try to get that fixed in a bit.

Reply:
That should fix it, nice catch!